### DISCUSSION TITLE
Config per Tenant

### DISCUSSION THREAD

**Discussion by josh9060** [vercel/next.js#53619] [Help]
### Summary

Hi all,

I have asked something similar in the Platforms repository. However, I think this concerns NextJS as a whole, and how something like this could be achieved pre app router (what we're using now) and with app router.

What do I mean by config per tenant? I will give our use case as an example. We use ORY Kratos - which by design when implementing multi-tenancy - will require each tenant to have a secret, specific redirect URL, etc. This is not new though, many multi-tenant applications may have tenant specific API keys, config, etc.

Handling a config per tenant on the backend is quite straightforward IMO (or rather has been solved many times). Think about establishing a DB connection for a tenant. You have a middleware that gets the tenant ID (be that from a token, request headers, etc) and put that into the request context. Bang. You can now switch DB connections per tenant. This is an over simplified example, but the solution is clear.

Now, take the same use case with NextJS. I want to return a callback URL to the client so we can log in. Sounds easy right? Well, it is... But, it gets messy. Say you return this value from SSP to your page component, but you use RQ, now your custom RQ hook must accept an argument for that value. It feels wrong. But not awful.

Now, what happens if we need to get a tenant specific secret key on an API route? We can, but first we must get the tenant ID and then fetch the tenant specific value.

This leads onto the final issue. Where do we store these tenant specific config options? Do the non sensitive values go in a DB and the sensitive value go into something like Hashicorp Vault? Or do we inject all of the into the env and prefix them with the tenant ID?

Just to be clear, I am not saying this is a NextJS issue and I am merely giving examples to illustrate an architectural problem I have.

So, how would you solve configuration and secrets per tenant? Would you just get secrets on every API call? Would you pass client values to the components and functions that need them? Or is there a better way?

Thanks for reading. I am most likely missing the best approach!

**Comment by WoetDev** [comment]
Wish this would've got an answer cause I'm wrestling with the same issue right now
